### PR TITLE
Initial rewrite of argument parsing with labels

### PR DESCRIPTION
**Pull request by casuallyblue**

### CODE REVIEW

**Comment by casuallyblue** [comment]
The implementation should work, but at the moment it seems that there's an issue that the labels make the compiler underflow one of the stacks when `ArgumentParser.new` gets called

**Comment by typeswitch-dev** [comment]
> The implementation should work, but at the moment it seems that there's an issue that the labels make the compiler underflow one of the stacks when `ArgumentParser.new` gets called

This was a problem before merging the data labels PR, I think I fixed it.
It looks like the PR compiled fine after merging from main. Have you tried running `make update`?

**Comment by casuallyblue** [comment]
that seems like it has fixed the issue

File: src/args/state.mth [hunk 1]
```diff
@@ -5,105 +5,108 @@ import(std.maybe)
...
 )
 
-def(ArgumentParserState.argv-info, ArgumentParserState(a) -- ArgumentParserStateArgvInfo,
-  ARGUMENT_PARSER_STATE -> drop3 swap drop)
+def(State.argv, State(a) -- List(Str), argv-info ArgvInfo.argv)
+
+def(State.program-name, State(a) -- Str, argv-info program-name)
+
+def(State.argv-info, State(a) -- ArgvInfo,
+  STATE -> argv-info> dup >argv-info STATE drop)
```
Comment (typeswitch-dev) [inline src/args/state.mth:71]: I'm working on generating these projection functions automatically :-)

**Comment by typeswitch-dev** [comment]
Thanks, this looks great!

**Pull request merged**